### DISCUSSION TITLE
add another project with examples

### DISCUSSION THREAD

**Discussion by glelouet** [phax/jcodemodel#91] [Show and tell]
followup to #88 

maybe get another repository dedicated to examples.

The rep would be 
jcodemodel-examples
 - pom.xml : type pom, dependency to jcodemodel version ${project.version}
 - example1
 -  - pom.xml : type jar, request to execute src/main/java/example/Compile which produces the .java into src/main/generated
 - example2
 - - pom.xml as above.

**Comment by glelouet** [comment]
ideally that would be a sub project of a multi module but then it could break things so for easier management, just create anotherproject. If you feel like it, later on you can create jcodemodel-root (pom project) that contains the jcodemodel and jcodemodel-examples sub modules.

**Reply by phax** [reply]
The best way to fiddle with the Maven module is to create a Maven Plugin.
So the layout would be like this:
1. jcodemodel
2. jcodemodel-example-maven-plugin - create Java code
3. jcodemodel-example
    * This one invokes the maven-plugin from step 2 into `target/generate-sources`
    * And compiles them afterwards

**Comment by glelouet** [comment]
what ? why ?

You just need to add in the pom a request to the exec plugin …
That's what I have in my poms : 
```

	<build>
		<plugins>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>exec-maven-plugin</artifactId>
				<version>1.6.0</version>
				<executions>
					<execution>
						<phase>generate-sources</phase>
						<goals>
							<goal>java</goal>
						</goals>
					</execution>
				</executions>
				<configuration>
					<mainClass>mycompiler.Compile</mainClass>
					<arguments>
						<argument>path to compile</argument>
						<argument>${project.basedir}/src/generated/java</argument>
					</arguments>
				</configuration>
			</plugin>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>build-helper-maven-plugin</artifactId>
				<version>1.7</version>
				<executions>
					<execution>
						<id>add-source</id>
						<phase>generate-sources</phase>
						<goals>
							<goal>add-source</goal>
						</goals>
						<configuration>
							<sources>
								<source>src/generated/java</source>
							</sources>
						</configuration>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
```

It tells maven two things : 

1. add the folder src/generated/java as source for compilation
2. during the generation of the source, call a specific class. In that case it should be a different phase because the class needs to be compiled first (so it should be between compile and verify ?).

The issue is that my class is actually in a jar as dependency so already compiled during the generate-sources phases, while local classes are not yet. I guess using the generate-test-classes phase (https://maven.apache.org/ref/3.6.3/maven-core/lifecycles.html#default_Lifecycle ) would be better , but then we should place the test classes in src/test/java ?
If that's the case no need to add the src/generated/java dir as source folder since src/test/java already is.

Maybe that would be enough : 
```
	<build>
		<plugins>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>exec-maven-plugin</artifactId>
				<version>1.6.0</version>
				<executions>
					<execution>
						<phase>generate-test-classes</phase>
						<goals>
							<goal>java</goal>
						</goals>
					</execution>
				</executions>
				<configuration>
					<mainClass>example2.Compile</mainClass>
					<arguments>
						<argument>${project.basedir}/src/test/java</argument>
					</arguments>
				</configuration>
			</plugin>
```

**Reply by phax** [reply]
Yes, that looks like a reasonable way forward. I looked at your compilation patch and I think it fits where nicely.

Unfortunately I am not (yet) familiar with the Java tooling API so I cannot fully estimate the consequences - one of them will be, that the compiling part no longer runs with a JRE but requires a JDK - am I right on that one? But due to the nice encapsulation, that should not impact existing JRE users....

But are't we mixing 2 objectives here: 
1. Creating examples that should work and are part of the documentation (which should also be committed to the repository)
2. Creating test code for your compiling proposal (which will nicely work on in-memory data structures)

They don't necessarily belong together. Ot is it just me being confused a bit?

**Comment by glelouet** [comment]
indeed I think there can be an issue if the JDK is not installed on the client OS.

**Comment by glelouet** [comment]
present in the last release :)
